add a DIalContext function to SpdyRoundTripper

tzneal opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Add a DIalContext function to SpdyRoundTripper and use it for creating underlying connections. Update remotecommand.RoundTripperFor to pass the rest.Config's Dial function to the SpdyRoundTripper.

Prior to this change, remotecommand.RoundTripperFor would ignore the Dial function set on the rest.Config.

#### Which issue(s) this PR fixes:

Fixes #112847

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

tzneal commented:
I found #109135 after coding this up and looking for an issue.  I had written some additional tests and that PR has merge conflicts, so I submitted this PR.

aojea commented:
can you split the changes in 2 different commits?
1. add the DialContext
2. plumb the DialFunc from the constructor

tzneal commented:
> 

Sure, two PRs or just two commits in the same PR?

aojea commented:
2 commits is fine

/cc @enj

aojea commented:
/ok-to-test

aojea commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go:
```diff
@@ -58,9 +58,13 @@ type SpdyRoundTripper struct {
 	conn net.Conn
 
-	// Dialer is the dialer used to connect.  Used if non-nil.
+	// Deprecated: Dialer is the dialer used to connect.  Used if non-nil and DialFuc is nil.
```
you mean `DialContext` is nil , no?

tzneal replied:
Yeah, I called it DialFunc originally.  I also considered DialWithContext, no strong preference on the name.

aojea commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go:
```diff
@@ -58,9 +58,13 @@ type SpdyRoundTripper struct {
+	// Deprecated: Dialer is the dialer used to connect.  Used if non-nil and DialFuc is nil.
 	Dialer *net.Dialer
 
+	// DialContext is the function used to create new underlying connections. Optional, if not specified a default
```
if not specified we use `Dialer`

aojea commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go:
```diff
@@ -199,6 +203,14 @@ func (s *SpdyRoundTripper) dialWithHttpProxy(req *http.Request, proxyURL *url.UR
 	return rwc, nil
 }
 
+type fnDialer struct {
```
is this only needed because `proxy.SOCKS5` has a Dialer interface as argument?

liggitt replied:
fwiw, proxy.SOCKS5 uses a context dialer if passed in, and returns an implementation that implements context dialing

https://github.com/kubernetes/kubernetes/blob/c56629803d4a34377110499a6e9985bd7d5b9d0c/vendor/golang.org/x/net/proxy/socks5.go#L17-L23

https://github.com/kubernetes/kubernetes/blob/c56629803d4a34377110499a6e9985bd7d5b9d0c/vendor/golang.org/x/net/internal/socks/socks.go#L151

Does this wrapping let the socks proxy dialer successfully recognize context dialing is supported here?

tzneal replied:
Good catch! fixing.

aojea commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go:
```diff
@@ -273,20 +287,43 @@ func (s *SpdyRoundTripper) tlsConn(ctx context.Context, rwc net.Conn, targetHost
+		return conn, nil
 	}
 
-	tlsDialer := tls.Dialer{
```
so, we basically copy the `dial` code from the tls stdlibrary https://cs.opensource.google/go/go/+/refs/tags/go1.19.2:src/crypto/tls/tls.go;l=120;drc=fdea8e21708e9ed04498a7f7b68cbee43005d58d

tzneal replied:
Yeah, I added a reference.

aojea commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go:
```diff
@@ -273,20 +287,46 @@ func (s *SpdyRoundTripper) tlsConn(ctx context.Context, rwc net.Conn, targetHost
-	dialer := s.Dialer
-	if dialer == nil {
-		dialer = &net.Dialer{}
+	dial := s.DialContext
+	if dial == nil && s.Dialer != nil {
+		dial = s.Dialer.DialContext
+	} else if dial == nil {
+		nd := &net.Dialer{}
+		dial = nd.DialContext
```
maybe this way is more intuitive?

```suggestion
	dial := s.DialContext
	if dial == nil {
	    if s.Dialer != nil {
		dial = s.Dialer.DialContext
        	} else {
		  nd := &net.Dialer{}
		dial = nd.DialContext
		}
	}
```

tzneal replied:
Seems fine, updating.

aojea commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go:
```diff
@@ -120,6 +121,24 @@ func localhostCertPool(t *testing.T) *x509.CertPool {
+	)
+	dialerUsed := dialerDefault
+	dialFunc := func(ctx context.Context, network, address string) (net.Conn, error) {
+		dialerUsed = dialerDialFunc
```
dialers should be thread safe, maybe you should use soe `atomic...`

tzneal replied:
In this case I don't think it matters as dialing occurs serially due to https://github.com/golang/go/blob/432158b69a50e292b625d08dcfacd0604acbabd3/src/net/dial.go#L451, however it doesn't hurt to add.

aojea commented:
/lgtm
/assign @enj @liggitt

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go:
```diff
@@ -273,20 +287,50 @@ func (s *SpdyRoundTripper) tlsConn(ctx context.Context, rwc net.Conn, targetHost
+		return conn, nil
+	}
+
+	// This code is identical to tls.Dial but replicated here so we can
+	// use our own DialContext function.
+	tlsConfig := s.tlsConfig
+	if tlsConfig == nil {
+		tlsConfig = &tls.Config{}
+	}
+
+	// If no ServerName is set, infer the ServerName
+	// from the hostname we're connecting to.
+	if tlsConfig.ServerName == "" {
+		tlsConfig = tlsConfig.Clone()
+		colonPos := strings.LastIndex(dialAddr, ":")
+		if colonPos == -1 {
+			colonPos = len(dialAddr)
+		}
+		tlsConfig.ServerName = dialAddr[:colonPos]
 	}
```
it's *really* unfortunate we have to duplicate this and manage handshaking ourselves... there's no way to use stdlib tls setup with a custom dialer?

tzneal replied:
Agreed, but I don't see a way around it given the current state of things:

```go
package tls

func Dial(network, addr string, config *Config) (*Conn, error) {
	return DialWithDialer(new(net.Dialer), network, addr, config)
}
func DialWithDialer(dialer *net.Dialer, network, addr string, config *Config) (*Conn, error) {
	return dial(context.Background(), dialer, network, addr, config)
}
func dial(ctx context.Context, netDialer *net.Dialer, network, addr string, config *Config) (*Conn, error) {
...
```

There's an issue https://github.com/golang/go/issues/9360 that suggests using a standard dialer interface throughout the Go stdlib when only dial methods are used.  I can't find any specific language that prohibits replacing a concrete type with an interface with respect to Go1 compatibility, but I suspect it's considered a breaking change as the issue is tagged v2.

liggitt commented on staging/src/k8s.io/client-go/transport/spdy/spdy.go:
```diff
@@ -48,6 +48,7 @@ func RoundTripperFor(config *restclient.Config) (http.RoundTripper, Upgrader, er
 		Proxier:    proxy,
```
should DialContext be added to RoundTripperConfig so we don't have to fix up the returned upgradeRoundTripper after the fact?

tzneal replied:
sgtm

aojea commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go:
```diff
@@ -273,20 +296,50 @@ func (s *SpdyRoundTripper) tlsConn(ctx context.Context, rwc net.Conn, targetHost
+		}
 	}
 
+	conn, err := dial(ctx, "tcp", dialAddr)
+	if err != nil {
+		return nil, err
+	}
 	if url.Scheme == "http" {
```
actually, this becomes the same (just adapted to this part) as

```
	if req.URL.Scheme == "https" {
		return s.tlsConn(req.Context(), proxyDialConn, targetHost)
	}
	return proxyDialConn, nil
```

we are already using s.tlsConn @liggitt 🤷

aojea replied:
we have to make sure that our usage of s.tlsConn never drifts from `tls.Dialer{}.DialContext`  🤔

aojea commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go:
```diff
@@ -273,20 +296,50 @@ func (s *SpdyRoundTripper) tlsConn(ctx context.Context, rwc net.Conn, targetHost
+		return conn, nil
+	}
+
+	// This code is identical to tls.Dial but replicated here so we can
```
can we reuse `s.tlsConn`?

tzneal committed:
client-go: pass dial function to the spdy upgrade 

 client-go: pass dial function to the spdy upgrade

This causes the spdy connection to be established with the dial
function configured in the rest config.

tzneal committed:
spdy: add a DialContext function to SpdyRoundTripper

Add a DialContext function to SpdyRoundTripper and use it for
creating underlying connections if supplied. Mark the exiting
Dialer as deprecated.

pr comments


- Implement the ContextDialer interface
- Add DialContext to spdy.RoundTripperConfig
asd

dims commented:
If you still need this PR then please rebase, if not, please close the PR
